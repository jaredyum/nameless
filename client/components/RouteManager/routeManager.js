import React from 'react';

// React router
import {
  HashRouter,
  Switch
} from 'react-router-dom';

// MUI Components
import { withStyles } from '@material-ui/core/styles';
import CssBaseline from '@material-ui/core/CssBaseline';

// Components
import RouteMiddleware from 'components/RouteMiddleware';
import Main from 'components/Main';
import Loading from 'components/Loading';
import ConnectedNav from 'components/ConnectedNav';

// Routes config
import ROUTES_CONFIG from 'routes';

// Styles
import styles from './styles';

class RouteManager extends React.Component {
  componentDidMount() {
    const { onAuthStateChange } = this.props;

    // Register the firebase auth listener.
    // NOTE: We only ever do this ONCE - Firebase will handle updates.
    onAuthStateChange();
  }

  render() {
    const {
      authed,
      hasRights,
      classes,
      notifications,
      currentNotification
    } = this.props;

    // The "null" state for authed and hasRights indicate that we are still
    // in the process of authorizaion/authentication.
    if (authed === null || hasRights === null) {
      return (
        <div className={classes.routeManager}>
          <Loading />
        </div>
      );
    }

    return (
      <HashRouter>
        <div className={classes.routeManager}>
          <CssBaseline />
          <ConnectedNav />
          <Main
            notifications={notifications}
            currentNotification={currentNotification}
          >
            <Switch>
              { ROUTES_CONFIG.map(routeProps => (
                <RouteMiddleware
                  key={routeProps.path || '404'}
                  authed={authed}
                  hasRights={hasRights}
                  {...routeProps}
                />
              ))}
            </Switch>
          </Main>
        </div>
      </HashRouter>
    );
  }
}

export const RouteManagerJest = RouteManager;

export default withStyles(styles)(RouteManager);
