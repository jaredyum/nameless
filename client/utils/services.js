/* eslint-disable no-console */
import axios from 'axios';

// Firebase config.
import firebase from 'services/firebase';

// Errors
import { API_ERRORS } from 'copy/Global/errors';

/**
 * Decorates the headers for a request.
 * @param {string} token The firebase token.
 * @return {!Object} The decorated headers object.
 */
const decorateHeaders = token => ({
  Authorization: `Bearer ${token}`,
  'Content-Type': 'application/json'
});

export const defaultCatch = (endPoint, source = 'services') => (e) => {
  console.error(`${source}/${endPoint}`, e);
  throw e;
};

/**
 * Extracts and parses the Firebase token to use for requests.
 * @return {!Promise} A promise resolving with the firebase token.
 */
export const getUserToken = () => firebase
  .then(auth => auth.currentUser)
  .then(user => user.getIdToken())
  .catch(defaultCatch('getUserToken'));

/**
 * Request wrapper for all request types (GET, POST, PUT, DELETE, etc).
 * @param {!Object} config The request configuration.
 * @param {string} config.url The URL for the request.
 * @param {string} config.method The method (GET, POST, etc) for the request.
 * @param {!Object} config.payload Optional request payload.
 * @return {!Promise} Promise containing the axios response.
 */
export const makeRequest = ({
  url,
  method = 'get',
  payload = null
}) => getUserToken()
  .then(token => ({
    headers: decorateHeaders(token),
    method,
    url,
    ...(
      payload ? { data: payload } : {}
    )
  }))
  .then(axios)
  .then((res) => {
    if (res && (res.status >= 200 || res.status < 400)) {
      return res;
    }
    throw new Error(API_ERRORS.BAD_REQUEST);
  })
  .then((res) => {
    if ('location' in res.headers) {
      return makeRequest(res.headers.location, 'GET', null, true);
    }

    return res;
  });
