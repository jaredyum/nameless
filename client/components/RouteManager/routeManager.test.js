// Test utils
import { testCommonComponentAttrs } from 'utils/tests';

import { RouteManagerJest } from './routeManager';

const defaultProps = {
  classes: {},
  onAuthStateChange: jest.fn(),
  authed: true,
  redirect: ''
};

describe('<RouteManager /> component', () => {
  testCommonComponentAttrs(RouteManagerJest, defaultProps);
});
