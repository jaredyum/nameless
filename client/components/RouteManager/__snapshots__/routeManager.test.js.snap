// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<RouteManager /> component when rendering should render a component matching the snapshot 1`] = `
<HashRouter>
  <div>
    <WithStyles(CssBaseline) />
    <withRouter(Connect(Navigation)) />
    <WithStyles(Main)>
      <Switch>
        <RouteMiddleware
          authed={true}
          component={[Function]}
          exact={true}
          key="/login"
          path="/login"
          unauthOnly={true}
        />
        <RouteMiddleware
          authed={true}
          component={[Function]}
          exact={true}
          key="/reset-password"
          path="/reset-password"
          unauthOnly={true}
        />
        <RouteMiddleware
          authed={true}
          component={[Function]}
          key="/(home)?"
          path="/(home)?"
        />
        <RouteMiddleware
          authed={true}
          component={[Function]}
          key="404"
        />
      </Switch>
    </WithStyles(Main)>
  </div>
</HashRouter>
`;
